Add pre-commit config

zxiiro opened the pull request:
Add a pre-commit config for devs using pre-commit as a lint
manager. Add some basic linters for now as a start.

Signed-off-by: Thanh Ha <zxiiro@gmail.com>

[Your PR description here]

## PR Checklist

- [ ] Wrote a good commit message & description [see guide below].
- [ ] Commit message starts with `[Resolve #issue-number]`.
- [ ] Added/Updated unit tests.
- [ ] Added/Updated integration tests (if applicable).
- [ ] All unit tests (`make test`) are passing.
- [ ] Used the same coding conventions as the rest of the project.
- [x] The new code passes flake8 (`make lint`) checks.
- [x] The PR relates to _only_ one subject with a clear title.
      and description in grammatically correct, complete sentences.

## Approver/Reviewer Checklist

- [ ] Before merge squash related commits.

## Other Information

[Guide to writing a good commit](http://chris.beams.io/posts/git-commit/)

zaro0508 commented:
i like this, nice start to linting with pre-commit. Was wondering if we can add something to the README or CONTRIBUTIONS docs for this?  something along the lines...

```
As a pre-deployment step we syntatically validate files
with [pre-commit](https://pre-commit.com).

Please [install pre-commit](https://pre-commit.com/#install) then run
`pre-commit install` to setup the git hooks.   Once configured the pre-commit
linters will automatically run on every git commit.  Alternatively you
can manually execute the validations by running `pre-commit run --all-files`.
```

zxiiro commented:
> i like this, nice start to linting with pre-commit. Was wondering if we can add something to the README or CONTRIBUTIONS docs for this? something along the lines...
> 
> ```
> As a pre-deployment step we syntatically validate files
> with [pre-commit](https://pre-commit.com).
> 
> Please [install pre-commit](https://pre-commit.com/#install) then run
> `pre-commit install` to setup the git hooks.   Once configured the pre-commit
> linters will automatically run on every git commit.  Alternatively you
> can manually execute the validations by running `pre-commit run --all-files`.
> ```

I added your snippet to CONTRIBUTING.md since end users shouldn't need to care about pre-commit details. I replaced the existing "make lint" with pre-commit since all make lint was doing was running flake8.

I think it would be worth in a later patch for us to go through CONTRIBUTING and update it entirely to make sure things detailed there are still relevant for us. If we're using pre-commit we can likely remove some of the Makefile configs too as I find launching things automatically with pre-commit a lot more convenient.
